### ISSUE TITLE
SchemaSpy hung at connecting relationships phase

### ISSUE THREAD

**Issue by uppalabharath** [schemaspy/schemaspy#250]
I am trying to generate the schema diagram for one of my databases but the operation always gets stuck at connecting relationships phase. 

## Steps to Reproduce (for bugs)
1. Use standard schemspy command to generate the schema diagram

## Context

* Database type (e.g. MySQL): Oracle 11g
* Database driver version (is you use a special one): classes12.jar
* Your DDL (sample how to reproduce the issue):
* Full command including arguments how you run SchemaSpy:
`java -jar schemaspy-6.0.0-rc2.jar -t orathin -dp ojdbc6.jar -db XXXXX -host XXXXX -port XXXXX -s XXXXX -u XXXXX -p XXXXX -cat catalog -o XXXXX`

## Your Environment

* Version used: SchemaSpy 6.0.0 rc2
* Operating System and version (e.g. Windows 10): Windows 10
* Java version: jdk 1.8.0_66

**Comment by npetzall** [comment]
How many dots do you get after "Connecting relationships"?
How many tables are there in the selected schema?
For how long do you wait before killing schemaspy?

**Comment by uppalabharath** [comment]
* The dots are around 6. 
* The tables are 3500
* I killed the schemaspy operation after 3 hours.

**Comment by npetzall** [comment]
Could you re-run and add following to end of command line 
```
--logging.pattern.console="%d{HH:mm:ss.SSS} %clr(%-5level) - %msg%n" --logging.level.org.schemaspy=TRACE
```
This will add trace logging and timestamp to each log output. This will break the progress output, just so you know. With trace it should keep printing in the console, when the printing stops it's hanged, if it starts printing same thing over and over again, probably a loop issue. Hopefully you don't have to run it for 3 hrs to get some sort of indication.

I looked at the code, I don't see anything obvious. Due to the number of tables, it might be some kind of memory issue it starts looping the GC or the connections are really complex.
Also the progress logging is rubbish, 3 dots are added when it starts, fourth is added just because and then it's one dot for each table except for the last one.

Just so I better understand how quickly did it get to "Connecting relationships", did it spend the most time in the Connecting relationships state.

**Comment by uppalabharath** [comment]
Connecting relationships....22:22:16.993 DEBUG - Connecting foreign keys to XXXXXXXX
.22:22:26.402 DEBUG - Connecting foreign keys to XXXXXXXXX
.22:22:26.407 DEBUG - Connecting foreign keys to XXXXXXXXX
22:22:28.430 DEBUG - Adding foreign key constraint XXXXXXXX

It hung at the above step.  

When the TRACE logger is off it took aroun 130 seconds to get the tables info.

**Comment by npetzall** [comment]
Am I correct never got past the first foreign key. And XXXXXX all references the same table.

**Comment by uppalabharath** [comment]
No 
 Connecting foreign keys to A
.22:22:26.402 DEBUG - Connecting foreign keys to B
.22:22:26.407 DEBUG - Connecting foreign keys to C
22:22:28.430 DEBUG - Adding foreign key constraint D to C

I have provided different characters to different tables,

**Comment by uppalabharath** [comment]
1. Is there a way where we can specify the memory arguments for the java process. 
2. I didn't find any gc logs rolling even in the TRACE mode enabled.

**Comment by npetzall** [comment]
Well it can connect its the adding that seems to be the problem.

As for memory it's just the normal heap -Xmx

Trace is only for our code, there are jvm settings for printing gc information..

If you have jdk you're always able to connect with jconsole or the prettier jmc or the open source visualvm.

**Comment by npetzall** [comment]
There also the running latest build and running the one from the release page.
In the Readme.md there is a link for latest/snapshot

**Comment by npetzall** [comment]
More recent java uses 1/4th of physical ram it seems. I haven't run any large databases recently so I'm unsure about the memory requirements. It still depends on number of tables, number of columns per table and so forth.

I've come a cross concurrency problems with HashMap.put but dosen't seem like multiple threads are involved the default dbThreads for orathin->ora is 1.

But a thread dump would be awesome.
https://access.redhat.com/solutions/19170 has some options.

**Comment by uppalabharath** [comment]
Let me re run and get the thread dump. 
Thanks

**Comment by uppalabharath** [comment]
Hi @npetzall , 

I am attaching the thread dumps. I didn't find any thing suspicious in the dumps. 

Thanks, 
Bharath. 
[schemaspy_tdump.txt](https://github.com/schemaspy/schemaspy/files/1737781/schemaspy_tdump.txt)
[schemaspy_tdump1.txt](https://github.com/schemaspy/schemaspy/files/1737782/schemaspy_tdump1.txt)
[schemaspy_tdump2.txt](https://github.com/schemaspy/schemaspy/files/1737783/schemaspy_tdump2.txt)
[schemaspy_tdump3.txt](https://github.com/schemaspy/schemaspy/files/1737784/schemaspy_tdump3.txt)
[schemaspy_tdump4.txt](https://github.com/schemaspy/schemaspy/files/1737785/schemaspy_tdump4.txt)
[schemaspy_tdump5.txt](https://github.com/schemaspy/schemaspy/files/1737786/schemaspy_tdump5.txt)

**Comment by rafalkasa** [comment]
@uppalabharath could you run the SchemaSpy on your database with `-noimplied` parameter.
Maybe implied relationship create the problem.

**Comment by npetzall** [comment]
-noimplied shouldn't, I say shouldn't make a difference.
Since it never returns from `databaseService.gatheringSchemaDetails(config, db, progressListener);`
and impliedEnabled is checked later when doing the htmlOutput.

I was working on the progressbar and everything worked so quickly with the testdatabase that had 4 tables. So never saw the progressbar in action. I created 1000 tables almost identical except for name and comment. Well it did hardly impact the XML generation. However the gathering schemadetails got really slow in my opinion.

So when I get the time I'll look into it, if it's lots of small queries to the database that does it or if it's some kind of lookup in the internal model that slowing things down. 

It's gonna be interesting to see if the processing per table is linear or exponential. Since I thought it was slow with 1000 tables, than it should be at least 3 times slower than that with 3500 tables.

**Comment by uppalabharath** [comment]
Hi @rafalkasa , 

Sure I shall try with parameter you mentioned. 

Thanks, 
Bharath

**Comment by uppalabharath** [comment]
[noimplied1.txt](https://github.com/schemaspy/schemaspy/files/1740292/noimplied1.txt)
[noimplied2.txt](https://github.com/schemaspy/schemaspy/files/1740293/noimplied2.txt)
[noimplied3.txt](https://github.com/schemaspy/schemaspy/files/1740294/noimplied3.txt)
[noimplied4.txt](https://github.com/schemaspy/schemaspy/files/1740295/noimplied4.txt)

Hi @rafalkasa , 

I tried with the noimplied parameter but not much luck. I am attaching the thread dumps for reference. 

Thanks, 
Bharath

**Comment by npetzall** [comment]
It will take 29 hrs to complete.

org.schemaspy.service.TableService#connectForeignKeys
```
        // also try to find all of the 'remote' tables in other schemas that
        // point to our primary keys (not necessary in the normal case
        // as we infer this from the opposite direction)
        if (table.getSchema() != null || table.getCatalog() != null) {
            try (ResultSet rs = db.getMetaData().getExportedKeys(table.getCatalog(), table.getSchema(), table.getName())) {
                // get the foreign keys that reference our primary keys
                // note that this can take an insane amount of time on Oracle (i.e. 30 secs per call)
```

**Comment by npetzall** [comment]
@uppalabharath since I know you contribute and have built the application.
could you possibly change
https://github.com/schemaspy/schemaspy/blob/dcf4a917fb3c0369ea793962a5256de124fc4534/src/main/java/org/schemaspy/service/TableService.java#L219

```
if (table.getSchema() != null || table.getCatalog() != null) {
```

to
```
if (false) {
```

Build it and try to run it against the database just to see if it appears not hung?
Run it with the `-debug` flag.

**Comment by uppalabharath** [comment]
Sure @npetzall 
Will try and let you know my findings

**Comment by uppalabharath** [comment]
Hi @npetzall , 

I made the change and tried and it worked fine. 
The diagram is generated. May I know the implications if I make this condition false. 

Thanks, 
Bharath

**Comment by npetzall** [comment]
From the comment it's seems to be a reverse lookup. Finding tables that has the current table as a reference.

But I need to check the documentation in DatabaseMetadata to be sure. 

As the comment says we most likely will find the as imported keys when we get to that table. Only when the table with the fk to the current isn't analyzed will you get a change in behavior. Cross schema references to be more precis. Or even cross catalog.

**Comment by npetzall** [comment]
#267 closed #290 merged this is best effort.

Only other possibility is to add custom query which can be implemented if needed.
